### PR TITLE
Res evap

### PR DESCRIPTION
**Pull request by oskarhurst**
## Problem Description

Describe the problem you are trying to solve.
Creation of resEvap 
## Solution

first draft as code for ResEvap Algorithm

## how you tested the change

Describe what was done to test the change. This section can be left blank 
if automated tests demonstrating usage are provided in the PR.

## Where the following done:

- [ ] Tests. Check all that apply:
   - [ ] Unit tests created or modified that run during ant test.
   - [ ] Integration tests created or modified that run during integration testing
         (Formerly called regression tests.)
   - [ ] Test procedure descriptions for manual testing
- [ ] Was relevant documentation updated?
- [ ] Were relevant config element (e.g. XML data) updated as appropriate

If you aren't sure leave unchecked and we will help guide you to want needs changing where.

### CODE REVIEW

File: src/main/java/decodes/tsdb/WaterTempProfiles.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        timeSeriesDAO = DAO;
+        startDepth = start;
+        increment = incr;
+        boolean loading = true;
+        double currentDepth = startDepth;
+        TimeSeriesIdentifier tsid;
+        tsid = timeSeriesDAO.getTimeSeriesIdentifier(wtpId);
+        while(loading){
+            try {
+                TimeSeriesIdentifier newtsid = tsid.copyNoKey();
```
Comment (MikeNeilson) [inline src/main/java/decodes/tsdb/WaterTempProfiles.java:53]: For consideration:

The cwms schema "profiles" use a separate time series to control depth of measurement, or height for atmospheric instead of embedding the value in the name. It's not really used yet so it makes sense that NWDM wasn't doing thing that way but it likely makes sense for this effort to follow that scheme. 

However as I type that out I'm not entirely sure how to describe things lining up other than having multiple measurements within the hour, which makes complete sense for actually measured profiles with the sensor moving but these are calculated all at once.


Keep this as is for now, I'll investigate the profiles more to see how it relates.

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+	/**
+	 * Algorithm-specific initialization provided by the subclass.
+	 */
+	protected void initAWAlgorithm( )
+		throws DbCompException
+	{
+//AW:INIT
+		//todo
+		//getParmRef("windSpeed").missingAction == MissingAction.IGNORE;
+        _awAlgoType = AWAlgoType.TIME_SLICE;
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:171]: Should be an `AWAlgoType.AGGREGATE`.

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 3]
```diff
@@ -0,0 +1,526 @@
...
+
+//AW:INIT_END
+
+//AW:USERINIT
+//AW:USERINIT_END
+	}
+
+	public TimeSeriesIdentifier makeTSID(String tsIdStr) throws DbIoException {
+		try {
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:180]: TimeseriesDb should have a function that already supports this behavior.

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 4]
```diff
@@ -0,0 +1,526 @@
...
+			}catch(NoConversionException ex){
+				throw new DbCompException("Failed to load value from timeseries"+ex);
+			}
+			i++;
+		}
+		setProfiles(DailyWTP, arrayWTP, newTime);
+	}
+
+	public void calcDailyEvap(NamedVariable output, CTimeSeries tsc,  Date CurrentTime) throws DbCompException, NoConversionException, DecodesException {
+		double TotalEvap = calcDailyACC(tsc, CurrentTime);
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:264]: since moving to aggregating algorithm you can just keep a running sum of the evap for the day, along with the number of values added to the sum.

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 5]
```diff
@@ -0,0 +1,526 @@
...
+		setOutput(HourlyLatent, computedList.get(2));
+		setOutput(HourlySolar, computedList.get(3));
+		setOutput(HourlyFluxIn, computedList.get(4));
+		setOutput(HourlyFluxOut, computedList.get(5));
+		setOutput(HourlyEvap, computedList.get(6));
+
+		setProfiles(hourlyWTP, resEvap.getHourlyWaterTempProfile(), _timeSliceBaseTime);
+
+		//TODO Aggregation
+		if(_timeSliceBaseTime.getDate() != LastDate.getDate()){
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:475]: Move to afterTimeSlice and you can ditch the date logic, with the aggregating algorithm you can be sure that the output will be put at the correct timestamp according to user specifications.

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 6]
```diff
@@ -0,0 +1,526 @@
...
+	/**
+	 * This method is called once after iterating all time slices.
+	 */
+	protected void afterTimeSlices()
+		throws DbCompException
+	{
+		hourlyWTP.SaveProfiles();
+		DailyWTP.SaveProfiles();
+
+		crd.close();
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:496]: don't forget the other Daos.

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 7]
```diff
@@ -0,0 +1,526 @@
...
+	
+	/**
+	 * This method is called once before iterating all time slices.
+	 */
+	protected void beforeTimeSlices()
+		throws DbCompException
+	{
+//AW:BEFORE_TIMESLICES
+		siteDAO =  tsdb.makeSiteDAO();
+		timeSeriesDAO = tsdb.makeTimeSeriesDAO();
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:336]: put the CwmsRatingDao here also.

**Review by MikeNeilson — CHANGES_REQUESTED** [review]
I'm going to leave it at that for now. There's several comments that are in the "to be done before merge" category and just here so I don't forget. Use your own judgement on when to perform any given step.

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 8]
```diff
@@ -221,72 +278,78 @@ public double[] getProfiles() throws Exception {
...
-			}
-			catch(Exception ex){
-				throw new DbCompException("Failed to load past Evap values"+ex);
-			}
-		}while(loopTime.getDate() == pastTime.getDate());
-
-		return total;
-	}
+	//Returns Converted double of cts from currUnits space to NewUnits
 	public double convertUnits(double cts, String currUnits, String newUnits) throws NoConversionException, DecodesException {
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:353]: ? The time series system contains a unit conversion method. 

NOTE: that's for units of the same abstract measurement (e.g. feet to meters). If this is doing something more than that the name "convertUnits" is insufficient.

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 9]
```diff
@@ -329,111 +400,142 @@ public double getMaxTempDepthMeters(){
...
+			reservoir.setName(reservoirId);
+			reservoir.setThermalDiffusivityCoefficient(ThermalDifCoe);
+			try {
+				reservoir.setWindShearMethod(WindShearMethod.fromString(WindShear));
+			} catch (Throwable ex) {
+				ex.printStackTrace();
+			}
 
+			reservoir.setInputDataIsEnglish(true);
+			double lonneg = -Longi; // why make longitude positive?
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:475]: Are we "correcting" user data here? In theory a negative Longitude could be correct for a future OpenDCS user.

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 10]
```diff
@@ -447,53 +549,66 @@ protected void beforeTimeSlices()
...
 	protected void doAWTimeSlice()
             throws DbCompException {
 //AW:TIMESLICE
-		try {
-			boolean noProblem = resEvap.compute(_timeSliceBaseTime, GMT_Offset);
-			if (!noProblem){
-				throw new DbCompException("ResEvap Compute Not Successful. Exiting Script.");
+		if (baseTimes.size() == 24) {
+			try {
+				boolean noProblem = resEvap.compute(_timeSliceBaseTime, GMT_Offset);
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:558]: Do we have a plan for better error handling than a boolean return?

File: src/main/java/decodes/cwms/algo/ResEvapAlgo.java [hunk 11]
```diff
@@ -447,53 +549,66 @@ protected void beforeTimeSlices()
...
-		if(_timeSliceBaseTime.getDate() != LastDate.getDate()){
-            try {
-                calcDailyEvap(DailyEvap, HourlyEvapTS, _timeSliceBaseTime);
-            } catch (NoConversionException | DecodesException e) {
-                throw new RuntimeException(e);
-            }
-            setDailyProfiles(_timeSliceBaseTime);
+			List<Double> computedList = resEvap.getComputedMetTimeSeries();
+			setOutput(HourlySurfaceTemp, computedList.get(0));
+			setOutput(HourlySensible, -computedList.get(1));
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/algo/ResEvapAlgo.java:568]: what's with the `-` sign?

File: src/main/java/decodes/cwms/resevapcalc/DownwellingInfraRedFlux.java [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    private static final double[][][][] COEF = new double[2][2][3][4];
+    static 
+    {
+        COEF[0][0] = COEF11;
+        COEF[0][1] = COEF12;
+        COEF[1][0] = COEF21;
+        COEF[1][1] = COEF22;
+    }
+    
+    public DownwellingInfraRedFlux()
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/resevapcalc/DownwellingInfraRedFlux.java:57]: This should probably be private. it appears everything in this class is static.

File: src/main/java/decodes/cwms/resevapcalc/EvapMetData.java [hunk 13]
```diff
@@ -8,74 +8,62 @@
...
+    CTimeSeries altitudeHighClouds;
     
     // these variables hold last valid values
     //was originally RMA undefinedDouble = -FloatMax
-    double _wsTemp_old = Constants.undefinedDouble;
-    double _windSpeed_old = Constants.undefinedDouble;
-    double _airTemp_old = Constants.undefinedDouble;
-    double _relHumidity_old = Constants.undefinedDouble;
-    double _airPressure_old = Constants.undefinedDouble;
+    double wsTemp_old = Constants.undefinedDouble;
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/resevapcalc/EvapMetData.java:37]: camel case variable names.

File: src/main/java/decodes/cwms/resevapcalc/EvapMetData.java [hunk 14]
```diff
@@ -8,74 +8,62 @@
...
+    double relHumidity_old = Constants.undefinedDouble;
+    double airPressure_old = Constants.undefinedDouble;
     
     // test variables hold the met values for the current time
-    double _wsTemp_current = Constants.undefinedDouble;
-    double _windSpeed_current = Constants.undefinedDouble;
-    double _airTemp_current = Constants.undefinedDouble;
-    double _relHumidity_current = Constants.undefinedDouble;
-    double _airPressure_current = Constants.undefinedDouble;
+    double wsTemp_current = Constants.undefinedDouble;
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/resevapcalc/EvapMetData.java:44]: camel case.

should probably expand ws, not sure what it means in this context (looking at nothing else.)

File: src/main/java/decodes/cwms/resevapcalc/EvapMetData.java [hunk 15]
```diff
@@ -8,74 +8,62 @@
...
-        return getMetValue( _windspeedTsc, hecTime);
+    public double getWindSpeed( Date hecTime ) throws ResEvapException {
+        return getMetValue(windspeedTsc, hecTime);
     }
     
-    public double getAirTemp( Date hecTime )
-    {
-        //return _airTempTsc.getValue(hecTime);
-        return getMetValue( _airTempTsc, hecTime);
+    public double getAirTemp( Date hecTime ) throws ResEvapException {
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/resevapcalc/EvapMetData.java:54]: hecTime is no longer the current name of this variable.

File: src/main/java/decodes/cwms/resevapcalc/EvapReservoir.java [hunk 16]
```diff
@@ -36,17 +39,19 @@
...
 	private static final Logger LOGGER = Logger.getLogger(EvapReservoir.class.getName());
     public static final int MAX_RES_AREAS = 50000;
     public static final int NLAYERS = 1000;
     public static final double FT_TO_M = 0.3048;
     public static final double ACRES_TO_M2 = 4046.8564224;
-    
+
+    public Connection conn;
     // package access to these variables
     public    int   _numberLayers;    //TODO, not current set
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/resevapcalc/EvapReservoir.java:51]: variables should not start with _

File: src/main/java/decodes/cwms/resevapcalc/ResEvap.java [hunk 17]
```diff
@@ -80,12 +60,12 @@ public class ResEvap
             
     // store Water temperature profile data
     // one profile for each hour
-    double[] _wtempProfiles;
+    double[] wtempProfiles;
             
-    public EvapReservoir _reservoir;
-    public EvapMetData _metData;
+    public EvapReservoir reservoir;
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/resevapcalc/ResEvap.java:65]: The naming of this class and EvapReservoir just seems... odd.

File: src/main/java/decodes/cwms/resevapcalc/ResEvap.java [hunk 18]
```diff
@@ -233,7 +203,7 @@ public boolean compute(Date currentTime,
             try
             {
             	xout = new BufferedWriter( new FileWriter( xoutfil ) );
-            	_reservoir.setDebugFile(xout);
+            	reservoir.setDebugFile(xout);
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/resevapcalc/ResEvap.java:206]: Any particular reason we don't just dump all this in the log?
We can use level checks to decide if something should actually log.

File: src/main/java/decodes/cwms/resevapcalc/ResWaterTemperatureCompute.java [hunk 19]
```diff
@@ -0,0 +1,646 @@
...
+        pe_mix_out = new double[EvapReservoir.NLAYERS];
+   }
+    
+    /**
+     * set output writer for diagnostic text output for reservoir 
+     * energy balance
+     *  
+     * @param tout
+     */
+    public void setOutfile( BufferedWriter tout )
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/resevapcalc/ResWaterTemperatureCompute.java:61]: This should probably be a more generic input. Just `Writer` the caller can then decide if they want it buffered or not.

File: src/main/java/decodes/cwms/resevapcalc/ResWaterTemperatureCompute.java [hunk 20]
```diff
@@ -0,0 +1,646 @@
...
+             
+ 
+        tempDepth = 0.25 ;
+        sum_energy_diff = 0.;
+    
+        for ( int i=resj; i>=0; i--)  //TODO check 
+        {
+            wt[i] = u[i];
+            rhow[i] = EvapUtilities.den_h2o(wt[i]);
+            rhow_org[i] = rhow[i];
```
Comment (MikeNeilson) [inline src/main/java/decodes/cwms/resevapcalc/ResWaterTemperatureCompute.java:360]: i would use orig or original. We tend to use org as short for organization.

File: src/main/java/decodes/tsdb/WaterTempProfiles.java (hunk 1 above)
Comment (MikeNeilson) [reply src/main/java/decodes/tsdb/WaterTempProfiles.java:56]: We established that these profiles are not the same as our "time series" profiles.

**Review by MikeNeilson — CHANGES_REQUESTED** [review]
Ignore the comments on the files a root, I just started there but it appears you just haven't deleted them yet.

The branch does need to be rebased onto main and re-targeted to main as well. We rearrange the folder structure slightly.


I'm going to have to do a deeper dive later, the general structure seems reasonable but there's enough diffs that it's hard to follow the code but wanted to give this feedback now.

**Comment by ktarbet** [comment]
closing this outdated PR in favor of a new PR: 
https://github.com/opendcs/opendcs/pull/791
